File: Recursion/lecture 10/Fraz_recursion_10.cpp
#include<bits/stdc++.h>
void helper(vector<int> &v, int i, vector<int> &subSet, vector<vector<int>> &ans)
{
    if (i == v.size())
    {
        ans.push_back(subSet);
        return;
    }
    // include the ith element
    subSet.push_back(v[i]);
    helper(v, i + 1, subSet, ans);
    // not including the ith element
    subSet.pop_back();
    while (i + 1 < v.size() && v[i] == v[i + 1])
        i++;
    helper(v, i + 1, subSet, ans);
}


vector<vector<int>> uniqueSubsets(int n, vector<int> &v)
{
    vector<vector<int>> ans;
    vector<int> subSet;
    // sorting before calling the function 
    sort(v.begin(),v.end());
    helper(v, 0, subSet, ans);
    sort(ans.begin(),ans.end());
    return ans;
}

File: Recursion/lecture 10/Fraz_recursion_10.js
const helper = (v, i, subSet, ans) => {
    // one of many ways of referencing object in JS
    let subSetRef = JSON.parse(JSON.stringify(subSet));
    
    if (i === v.length)
    {
        ans.push(subSetRef);
        return;
    }
    // include the ith element
    
    subSetRef.push(v[i]);
    helper(v, i + 1, subSetRef, ans);
    // not including the ith element
    subSetRef.pop();
    while (i + 1 < v.length && v[i] === v[i + 1])
    {
        i++;
    }
    helper(v, i + 1, subSetRef, ans);
}
var pwset = (v) => {
    let ans = [];
    let subSet = [];
    helper(v, 0, subSet, ans);
    ans.sort();
    return ans;
};